fix(helm): do not use local chart if --repo is provided

jackwilsdon opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
If `--repo` is passed to `helm install` or `helm upgrade`, Helm will currently prefer a local directory with the name of the chart instead of the chart from the provided repository. This PR changes this behaviour such that the local filesystem is not checked when `--repo` is provided.

Fixes #11141

**Special notes for your reviewer**:
I considered keeping the current behaviour if the name is an absolute or relative path (i.e. it starts with `/` or `.`), but it doesn't seem correct to use a local chart if `--repo` is passed.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

gjenkins8 reviewed (commented):
Not a maintainer, but I think unit tests would be nice/needed to show that the behavior works as expected (and continues to work in the future)

jackwilsdon commented:
Added a test verifying that a remote chart is used when `--repo` is passed. I tried adding a golden output file, but there's a dyamic port number in the error message making it not really feasible.

joejulian commented:
This changes the default behavior. Changes of behavior are precluded without opting in by [hip-0004](https://github.com/helm/community/blob/main/hips/hip-0004.md).

joejulian commented:
Maybe use a different flag? `--remote-repo` for example

jackwilsdon committed:
fix(helm): introduce --remote-repo flag

Ignore local directories with the same name as the chart being installed
if --remote-repo is passed.

Fixes #11141

Signed-off-by: Jack Wilsdon <jack+github@wilsdon.me>

jackwilsdon commented:
Introduced a new `--force-remote` flag :+1:

joejulian commented:
Rather than repeat this all, please consider this comment as well.
https://github.com/helm/helm/pull/7874#issuecomment-1674062156

jackwilsdon commented:
Does it make sense to decide between `LocateChart` and `LocateRemoteChart` at the call site when `RemoteRepo` is a flag on `ChartPathOptions`? It seems like the logic would belong in `LocateChart`, as that function is on the options struct itself.

In terms of API users, they can also set the `RemoteRepo` flag when calling `LocateChart` if they only want to locate remote charts.

If it does make sense then I'm happy to implement it this way :+1:

joejulian commented:
I think the reason not to choose ahead of time is that it doesn't require anyone using the API to substantively change their code. By simply changing the option, they can get the new behavior.

In fact, now that I think of it, I'd not export locateRemoteChart so the use of the API is clear.

jackwilsdon commented:
Sounds good, I've pushed the suggested changes.
